r-package/grf/R/rank_average_treatment.R

```diff
@@ -0,0 +1,310 @@
+  if (ncol(priorities) > 1) {
+    point.estimate <- cbind(point.estimate, point.estimate[, 1] - point.estimate[, 2])
+    std.errors <- cbind(std.errors, apply(boot.output[["t"]][,, 1] - boot.output[["t"]][,, 2], 2, sd))
+  }
+  point.estimate[abs(point.estimate) < 1e-15] <- 0
+  std.errors[abs(std.errors) < 1e-15] <- 0
+  # ensure invariance: always >= 0
+  if (R < 2) {
+    std.errors[] <- 0
+  }
+  priority <- c(colnames(priorities), paste(colnames(priorities), collapse = " - "))[1:length(point.estimate[1, ])]
+
+  output <- list()
+  class(output) <- "rank_average_treatment_effect"
+  output[["estimate"]] <- point.estimate[1, ]
+  output[["std.err"]] <- std.errors[1, ]
+  output[["target"]] <- paste(priority, "|", target)
+  output[["TOC"]] <- data.frame(estimate = c(point.estimate[-1, ]),
+                                std.err = c(std.errors[-1, ]),
+                                q = q,
+                                priority = priority[gl(length(priority), length(q))],
+                                stringsAsFactors = FALSE)
+
+  output
+}
+
+#' Simple clustered bootstrap.
+#'
+#' Inspired by the `boot` function in the bootstrap package with clusters + half-sampling added.
+#' A future TODO could be to add parallel (not necessarily worth it)
+#' https://stat.ethz.ch/R-manual/R-devel/library/parallel/doc/parallel.pdf
+#'
+#' @param data A data frame with the original data.
+#' @param statistic A function computing estimate(s) with signature (data, indices, ...) where
+#' data is the original data, and indices a vector which defines the bootstrap sample.
+#' @param R The number of bootstrap replications.
+#' @param clusters Integer vector of cluster assignment, setting to 1:N corresponds to an ordinary
+#'  unclustered bootstrap.
+#' @param half.sample Whether to do half sample bootstrap (half the clusters are drawn). Default is TRUE.
+#' @param ... Additional arguments passed on to statistic.
+#' @return A list with the original estimate t0, and bootstrap estimates t.
+#'
+#' @references Angelo Canty and Brian Ripley (2021). boot: Bootstrap R (S-Plus) Functions.
+#'
+#' @keywords internal
+boot_grf <- function(data, statistic, R, clusters, half.sample = TRUE, ...) {
+  samples.by.cluster <- split(seq_along(clusters), clusters)
+  n <- length(samples.by.cluster) # number of clusters
+  if (n <= 1 || (half.sample && floor(n / 2) <= 1)) {
+    stop("Cannot bootstrap sample with only one effective unit.")
+  }
+  if (half.sample) {
+    n.bs <- floor(n / 2)
+    index.list <- replicate(R, unlist(samples.by.cluster[sample.int(n, n.bs, replace = FALSE)], use.names = FALSE), simplify = FALSE)
+  } else {
+    index.list <- replicate(R, unlist(samples.by.cluster[sample.int(n, replace = TRUE)], use.names = FALSE), simplify = FALSE)
+  }
+
+  t0 <- statistic(data, seq_len(NROW(data)), ...)
+  t0 <- matrix(unlist(t0), ncol = length(t0))
+
+  res <- lapply(seq_len(R), function(i) statistic(data, index.list[[i]], ...))
```

Comment:
This line can be parallelized with a drop-in replacement: `parallell::mclapply` instead of `lapply`

I am fine keeping it serial though, a quick bench shows it's reasonably fast, for ex n=1e6, continuous scoring and R=200 bs replications takes around 50-60 seconds, while mclapply with 24 cores barely bumps it down to ~25-30 seconds

I don't think that speedup is worth the hazzle of having to deal with `parallell` in the GRF package (even if it defaults to 1 core): it doesn't work on Windows, it relies on forking, which may cause problems in Rstudio, etc